Add canvas file read/write API

newsch opened the pull request:
Goal: Add functionality to save canvases and load canvases from a `.txt` file with line endings.

Closes #12.

TODO:
- [x] rebase off of new master
- [x] add function to read canvas from string with line endings
- [x] add functions to read and write canvases from files
- [x] functions have docstrings
- [x] functions have tests/have been tested
- [x] fix trim bugs (reimplement as canvas_trim?)

newsch committed:
Add more options for loading strings

newsch committed:
Initialize canvases with spaces

TODO: write a test

newsch committed:
Add canvas file read/write pseudocode

newsch committed:
Working reads from file

newsch committed:
Add a quick and dirty resize function

newsch committed:
Connect to frontend.c

newsch committed:
Fix canvas_fprint bug

newsch commented:
There's a working implementation of reading from and saving to files now.

Rebased off of `adam/implement_resizing`, because that has the working canvas-as-a-backend.

The canvas size is currently hard-coded at 1000x1000, and some of the api functions are still pretty ugly.

newsch commented:
Rebased off of the newly-rebased `adam/implement_resizing` (#18).

newsch commented:
Rebased off of `master`.

newsch commented:
I've got this working with the new `master` now. I made some design decisions that I'd like feedback on.

newsch committed:
Remove old debug file

newsch commented on src/frontend.c:
```diff
@@ -111,7 +140,18 @@ int main(int argc, char *argv[]) {
   State *state = &new_state;
 
-  while ((state->ch_in = wgetch(canvas_win))) {
+  while (1) {
```
Rather than make a "file read/write" mode, I added another layer that intercepts keypresses before they're sent to the modes, which lets you use these commands in any mode.

I'm not sure about the function name or the method of passing in the window vs a read character. Thoughts?

newsch replied:
No longer relevant - overhauled in #45

newsch commented on src/frontend.c:
```diff
@@ -56,6 +60,32 @@ int main(int argc, char *argv[]) {
 #endif
   logd("Starting frontend\n");
 
+  // init canvas from file
```
I'd like to get `stdin` working so you can pipe input from other programs to init a canvas. I'm running into some issues right now though.

newsch committed:
Add min and max macros

newsch committed:
Add canvas subset copy function

newsch committed:
Fill out canvas API

Added functions with tests:
- load one canvas into another canvas at a point, with an optional
  transparent character
- check if a point or index is inside a canvas

newsch committed:
Add canvas API docstrings, remove commented code

newsch committed:
Comment out unused variable

newsch committed:
Trim canvas trailing whitespace when saving

newsch committed:
Trim canvas trailing newlines when saving

newsch committed:
Add trim function to frontend

newsch commented:
Trimming!
[![asciicast](https://asciinema.org/a/j4IwRLDfZJBhhtXpfpLPSApnc.svg)](https://asciinema.org/a/j4IwRLDfZJBhhtXpfpLPSApnc)

I think this is good to go.

newsch committed:
Update canvas_resize to support shrinking

newsch committed:
Reorganize canvas.h

newsch committed:
Implement canvas_readf without rewind, working read from stdin

newsch commented:
Got reading from `stdin` working, it was a mix of using `rewind` and receiving an `EOF` from the input that seemed to do it.

[![asciicast](https://asciinema.org/a/icwGwEqEzVWRySCuwKTH2Erla.svg)](https://asciinema.org/a/icwGwEqEzVWRySCuwKTH2Erla)

MatthewBeaudouinLafon commented on src/frontend.c:
```diff
@@ -126,6 +169,77 @@ int main(int argc, char *argv[]) {
+int global_handler(WINDOW *win) {
+  int c = wgetch(win);
+  switch (c) {
+    case 18:  // ^R, read from savefile
```
Try the following for cleaner switch statement 

```
#ifndef CTRL
#define CTRL(c) ((c) & 037)
#endif

switch(getch()) {
case CTRL('r'):
    /* key ctrl-r (i.e. ^R) pressed */
```

newsch replied:
Added to `util.h`

MatthewBeaudouinLafon reviewed (commented):
We also talked about using helper functions in modes to handle semi-generic things like arrow movement such that the mode can overwrite it.

newsch commented:
The trimming is borked right now - some lines get switched around when the hoarded newlines are printed. Rather than fix it here, I'll factor it out into a separate canvas function, which should look a little cleaner.

newsch committed:
Integrate file saving and reading with frontend

newsch committed:
Don't trim trailing newlines

newsch committed:
Add more project-related files to gitignore

newsch committed:
Fix trimc and cpy_p1p2 bugs

newsch committed:
Fix trimc bug

I tried to make the previous version efficient by only checking up
to the limits it had found, but that meant it missed lower vertical
limits that were within the horizontal ones.

newsch committed:
Fix unused variable warning

labseven reviewed (approved):
Awesome stuff :+1:

labseven commented on src/fe_modes.c:
```diff
@@ -105,29 +177,51 @@ int free_line_arrows_to_char(int last_arrow, int current_arrow) {
+    num_left -= num_to_write;
+  }
 
+  // print_status("foo");
```
Remove unused lines

newsch commented:
Rebased off of new master
